go/client/chat_utils.go

```diff
@@ -32,3 +39,137 @@ func (m uidUsernameMapper) getUsername(ctx context.Context, g *libkb.GlobalConte
+			Value: "chat",
+			Usage: `Specify topic name of the conversation. Has to be chat or dev`,
+		},
+	}...)
+}
+
+type conversationResolver struct {
+	TlfName   string
+	TopicName string
+	TopicType chat1.TopicType
+}
+
+func (r conversationResolver) Resolve(ctx context.Context, chatClient keybase1.ChatLocalInterface) (ids []chat1.ConversationID, err error) {
+	ids, err = chatClient.ResolveConversationLocal(ctx, keybase1.ResolveConversationLocalArg{
+		TlfName:   r.TlfName,
+		TopicName: r.TopicName,
+		TopicType: r.TopicType,
+	})
+	return ids, err
+}
+
+type messageFetcher struct {
+	selector keybase1.MessageSelector
+	resolver conversationResolver
+
+	chatClient keybase1.ChatLocalInterface // for testing only
+}
+
+func makeMessageFetcherFromCliCtx(ctx *cli.Context, tlfName string, markAsRead bool) (fetcher messageFetcher, err error) {
+	fetcher.selector.MessageTypes = []chat1.MessageType{chat1.MessageType_TEXT, chat1.MessageType_ATTACHMENT}
+	fetcher.selector.Limit = ctx.Int("number")
+
+	if timeStr := ctx.String("time"); len(timeStr) > 0 {
+		fetcher.selector.Since = &timeStr
+	}
+
+	if ctx.Bool("all") {
+		fetcher.selector.Limit = 0
+	}
+
+	fetcher.resolver.TopicName = ctx.String("topic-name")
+	switch t := strings.ToLower(ctx.String("topic-type")); t {
+	case "chat":
+		fetcher.resolver.TopicType = chat1.TopicType_CHAT
+	case "dev":
+		fetcher.resolver.TopicType = chat1.TopicType_DEV
+	default:
+		err = fmt.Errorf("invalid topic-type %s. Has to be one of %v", t, []string{"chat", "dev"})
+		return fetcher, err
+	}
+
+	fetcher.selector.MarkAsRead = markAsRead
+	fetcher.resolver.TlfName = tlfName
+
+	return fetcher, nil
+}
+
+func (f messageFetcher) fetch(ctx context.Context, g *libkb.GlobalContext) (messages cliChatMessages, err error) {
+	chatClient := f.chatClient // should be nil unless in test
+	if chatClient == nil {
+		chatClient, err = GetChatLocalClient(g)
+		if err != nil {
+			return nil, fmt.Errorf("Getting chat service client error: %s", err)
+		}
+	}
+
+	conversationIDs, err := f.resolver.Resolve(ctx, chatClient)
+	if err != nil {
+		return nil, err
+	}
+	// TODO: prompt user to choose conversation(s)
+	f.selector.Conversations = conversationIDs
+
+	var mapper uidUsernameMapper
+	conversations, err := chatClient.GetMessagesLocal(ctx, f.selector)
+	if err != nil {
+		return nil, fmt.Errorf("GetMessagesLocal error: %s", err)
+	}
+
+	for _, conv := range conversations {
+		if len(conv.Messages) == 0 {
+			continue
+		}
+		m := conv.Messages[0]
+		var body string
+		switch t := m.MessagePlaintext.MessageBodies[0].Type; t {
+		case chat1.MessageType_TEXT:
+			body = formatChatText(m.MessagePlaintext.MessageBodies[0].Text)
+		case chat1.MessageType_ATTACHMENT:
+			body = formatChatAttachment(m.MessagePlaintext.MessageBodies[0].Attachment)
+		default:
+			g.Log.Debug("unsurported MessageType: %s", t)
+			continue
+		}
+
+		username, err := mapper.getUsername(ctx, g, keybase1.UID(m.MessagePlaintext.ClientHeader.Sender.String()))
```

Comment:
The mapper does use service (through LoadUser) to get username, but just caches the result in the CLI process. This is to avoid hitting the service for every single message when printing a list of messages. Do you want me to move this to `libkb`?